src/colmap/util/logging.h

```diff
@@ -101,4 +139,86 @@ bool __CheckOptionOpImpl(const char* file,
   }
 }
 
+template <typename T>
+inline std::string ToString(T msg) {
+  return std::to_string(msg);
+}
+
+inline std::string ToString(std::string msg) { return msg; }
+
+inline std::string ToString(const char* msg) { return std::string(msg); }
+
+template <typename T>
+inline T TemplateException(const char* file,
+                           const int line,
+                           const std::string& txt) {
+  std::stringstream ss;
+  ss << "[" << __GetConstFileBaseName(file) << ":" << line << "] " << txt;
+  // TODO: cleanup
+  return T(ss.str());
+}
+
+inline std::string __GetCheckString(const char* cond_str) {
+  // TODO: cleanup
+  std::stringstream ss;
+  ss << "Check Failed: " << cond_str;
+  return ss.str();
+}
+
+inline void __ThrowCheckImpl(const char* file,
+                             const int line,
+                             const bool result,
+                             const char* expr_str) {
+  if (!result) {
+    throw TemplateException<std::invalid_argument>(
+        file, line, __GetCheckString(expr_str).c_str());
+  }
+}
+
+inline void __ThrowCheckImplMsg(const char* file,
+                                const int line,
+                                const bool result,
+                                const char* expr_str,
+                                const std::string& msg) {
+  if (!result) {
+    std::string m = std::string(expr_str) + " : " + msg;
+    throw TemplateException<std::invalid_argument>(
+        file, line, __GetCheckString(m.c_str()));
+  }
+}
+
+template <typename T>
+T __ThrowCheckNotNull(const char* file,
+                      const int line,
+                      T&& t,
+                      const char* name) {
+  if (t == nullptr) {
+    std::string msg = "\"" + std::string(name) + "\" Must be non nullptr";
+    throw TemplateException<std::invalid_argument>(
+        file, line, __GetCheckString(msg.c_str()));
+  } else {
+    return std::forward<T>(t);
+  }
+}
+
+template <typename T1, typename T2>
+void __ThrowCheckOpImpl(const char* file,
+                        const int line,
+                        const bool result,
+                        const T1& val1,
+                        const T2& val2,
+                        const char* val1_str,
+                        const char* val2_str,
+                        const char* op_str) {
+  if (!result) {
+    // TODO: cleanup
```

Comment:
There are various cleanup TODOs in the new code. What is the specific cleanup that is still left?

Reply:
I've replaced the string streams with string concatenation, which should be faster.